File: solutions/shuffle-an-array/shuffle-an-array.py
import random

class Solution:

    def __init__(self, nums: List[int]):
        self.nums = nums

    def reset(self) -> List[int]:
        return self.nums
        

    def shuffle(self) -> List[int]:
        if len(self.nums) < 2:
            return self.nums
        
        copy = self.nums[0:]
        # python passes lists by reference so create copy with slice
        result = []
        
        while len(copy) > 0:
            length = len(copy) - 1
            index = random.randint(0, length)
            result.append(copy[index])
            copy.remove(copy[index])
        
        return result


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()

File: solutions/shuffle-an-array/shuffle-an-array.js
/**
 * @param {number[]} nums
 */
var Solution = function(nums) {
    this.initial = nums;
    this.random = [];
};

/**
 * Resets the array to its original configuration and return it.
 * @return {number[]}
 */
Solution.prototype.reset = function() {
    return this.initial;
};

/**
 * Returns a random shuffling of the array.
 * @return {number[]}
 */
// get a copy of the initial array
// create container
// loop while copy has items
  // get a random index
  // push item with index to result
  // remove item from copy array
// return result

Solution.prototype.shuffle = function() {
    let nums = this.initial.slice();
    let result = [];
    
    while (nums.length > 0) {
        let index = Math.floor(Math.random() * nums.length);
        result.push(nums[index]);
        nums.splice(index, 1);
    }
    
    return result;
};

/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(nums)
 * var param_1 = obj.reset()
 * var param_2 = obj.shuffle()
 */
